rpc/connection_log.go

```diff
@@ -0,0 +1,155 @@
+		logPrefix: strings.Join(
+			[]string{prefix, hex.EncodeToString(randBytes)}, " "),
+	}
+}
+
+func formatLogFields(f string, lf ...logField) string {
+	fields := make([]interface{}, 0, len(lf))
+	for _, lf := range lf {
+		fields = append(fields, lf)
+	}
+	return fmt.Sprintf(f, fields...)
+}
+
+func (l *connectionLogUnstructured) Warning(
+	format string, fields ...logField) {
+	l.LogOutput.Warning("(%s) %s", l.logPrefix,
+		formatLogFields(format, fields...))
+}
+
+func (l *connectionLogUnstructured) Debug(
+	format string, fields ...logField) {
+	l.LogOutput.Debug("(%s) %s", l.logPrefix,
+		formatLogFields(format, fields...))
+}
+
+func (l *connectionLogUnstructured) Info(
+	format string, fields ...logField) {
+	l.LogOutput.Info("(%s) %s", l.logPrefix,
+		formatLogFields(format, fields...))
+}
+
+// LogrusEntry and LogrusLogger define methods we need from logrus to avoid
+// pulling logrus as dependency.
+type LogrusEntry interface {
+	Debug(args ...interface{})
+	Info(args ...interface{})
+	Warning(args ...interface{})
+}
+
+// LogrusLogger maps to *logrus.Logger, but will need an adapter that converts
+// map[string]interface{} to logrus.Fields, and adapt logrus.Entry to
+// LogrusEntry.
+type LogrusLogger interface {
+	WithFields(map[string]interface{}) LogrusEntry
+}
+
+type connectionLogLogrus struct {
+	log LogrusLogger
+
+	section   string
+	randBytes string
+}
+
+func newConnectionLogLogrus(
+	log LogrusLogger, section string) *connectionLogLogrus {
+	randBytes := make([]byte, 4)
+	rand.Read(randBytes)
+	return &connectionLogLogrus{
+		log:       log,
+		section:   section,
+		randBytes: hex.EncodeToString(randBytes),
+	}
+}
+
+func (l *connectionLogLogrus) getLogEntryAndExtractMsg(
+	format string, fields []logField) (entry LogrusEntry, msg string) {
+	_, file, line, ok := runtime.Caller(2)
+	if !ok {
+		file, line = "unknown", -1
+	}
+	mFields := make(map[string]interface{})
+	for _, f := range fields {
+		mFields[f.key] = f.value
+	}
+	mFields["section"] = l.section
+	mFields["identifier"] = l.randBytes
+	mFields["file"], mFields["line"] = file, line
+
+	if msgI, ok := mFields[msgKey]; ok {
+		if msg, ok = msgI.(string); !ok {
+			msg = fmt.Sprintf("%s", msgI)
+		}
+	} else {
+		// If "msg" isn't present in fields, just use a formatted string. It
+		// won't index well but it's better than losing information.
+		msg = formatLogFields(format, fields...)
+	}
+
+	return l.log.WithFields(mFields), msg
+}
+
+func (l *connectionLogLogrus) Warning(format string, fields ...logField) {
+	e, msg := l.getLogEntryAndExtractMsg(format, fields)
+	e.Warning(msg)
+}
+
+func (l *connectionLogLogrus) Debug(format string, fields ...logField) {
+	// Nasty hack to lift Debug to Info: we can't just do Info everywhere in
+	// this package since they'll leak to CLI in keybase/client. We also can't
+	// just turn on Debug logging on server-side. Since logrus loggers are only
+	// used by server side, we simply override Debug with Info here.
```

Comment:
Why can't we just enable debug logging on the servers? I really appreciate that you documented this approach, but I don't agree that we should do it.

Reply:
I think we turned off debug logging earlier because there was too much logging? I don't remember exactly though.

IMO these connection.go logging really should be info. They only happen when the connection state changes, which is pretty rare for server<->server communications.

Reply:
I'm not sure whether you're agreeing or disagreeing with me...

Reply:
Ah; I just don't know what the answer is regarding to log level setting with kbfs-server. But I guess if we decide to switch log level back debug, it should be another ticket and I can add a TODO here about removing this hack if we ever do so.

Reply:
No, I don't think we should merge with this hack at all.

Reply:
Hmmm sounds like the only way to move forward with this PR is to remove all the existing Debug logging in `kbfs-server` and turn on Debug logging from `keybase/aws` then?